app/plugin_install.go

```diff
@@ -16,12 +17,31 @@ import (
 	"github.com/mattermost/mattermost-server/utils"
 )
 
+const managedPluginFileName = ".filestore"
+const fileStorePluginFolder = "./plugins"
+
 // InstallPlugin unpacks and installs a plugin but does not enable or activate it.
 func (a *App) InstallPlugin(pluginFile io.ReadSeeker, replace bool) (*model.Manifest, *model.AppError) {
 	return a.installPlugin(pluginFile, replace)
 }
 
 func (a *App) installPlugin(pluginFile io.ReadSeeker, replace bool) (*model.Manifest, *model.AppError) {
+	manifest, installErr := a.installPluginLocally(pluginFile, replace)
+	if installErr != nil {
+		return nil, installErr
+	}
+
+	// Store bundle in the file store to allow access from other servers.
+	pluginFile.Seek(0, 0)
+
+	if _, err := a.WriteFile(pluginFile, a.getBundleStorePath(manifest.Id)); err != nil {
+		return nil, model.NewAppError("uploadPlugin", "app.plugin.store_bundle.app_error", nil, err.Error(), http.StatusInternalServerError)
```

Comment:
was just wondering what should happen if we can't store the plugin in the filestore, should we remove it locally because next time the server starts, it would be removed, no?

Reply:
Yes, but it does return an error to the client app about the problem regarding the file store. Given that it will be removed on restart later, removing it locally would leave the user blocked on installing it until they try again or fix their problem with the file store. Thoughts here on what would be the best use case ? @lieut-data @aaronrothschild

Reply:
Is it correct to assume the plugin won't start until after we upload to the filestore? If so, I'm inclined to leave the state as-is, viewing `plugins` as a sort of `tmp` directory that will get cleaned up later.

Reply:
The plugin is installed locally first and, if successful it is uploaded to the filestore. By default it is installed disabled, so the plugin will not start until the user configures and enables it.

Reply:
@marianunez but is it the case that you can enable it after? even though the plugin is not stored in the filestore?

Reply:
@ali-farooq0 correct. It will be available temporary until next time the server is restarted or they reinstall it again after correcting any file store issue that caused the problem.

Reply:
Ah, then in that case I think I understand better @ali-farooq0's proposal to (try to) remove the local plugin if we can't upload to the file store. Thoughts?

Reply:
To be consistent, then we should also reinstall when it fails to remove it from the filestore? So, there would be rollback operation if anything fails in the file store?

Maybe we can move rollback operations for future improvements if we find it's a better user experience?

Reply:
would it not automatically reinstall on server-restart when we fail to remove from FS?

The only worry I have with not removing it here is that on server-restart the plugin will automatically disappear causing some confusion to the admin. However since this is an edge case maybe we can log a jira and fix it as a future improvement as @marianunez suggested.

Thoughts @lieut-data?